sound/soc/sof/intel/hda-loader.c

```diff
@@ -315,55 +317,54 @@ int hda_dsp_cl_boot_firmware(struct snd_sof_dev *sdev)
 {
 	struct snd_sof_pdata *plat_data = dev_get_platdata(sdev->dev);
 	struct firmware stripped_firmware;
-	int ret, tag;
+	int ret, tag, i;
 
 	stripped_firmware.data = plat_data->fw->data;
 	stripped_firmware.size = plat_data->fw->size;
 
-	tag = cl_dsp_init(sdev, stripped_firmware.data,
-			  stripped_firmware.size);
-
-	/* retry enabling core and ROM load. seemed to help */
-	if (tag < 0) {
+	/* try attempting fw boot a few times before giving up */
+	for (i = 0; i < HDA_FW_BOOT_ATTEMPTS; i++) {
 		tag = cl_dsp_init(sdev, stripped_firmware.data,
 				  stripped_firmware.size);
+
 		if (tag <= 0) {
 			dev_err(sdev->dev, "Error code=0x%x: FW status=0x%x\n",
 				snd_sof_dsp_read(sdev, HDA_DSP_BAR,
 						 HDA_DSP_SRAM_REG_ROM_ERROR),
 				snd_sof_dsp_read(sdev, HDA_DSP_BAR,
 						 HDA_DSP_SRAM_REG_ROM_STATUS));
-			dev_err(sdev->dev, "Core En/ROM load fail:%d\n",
-				tag);
+			dev_err(sdev->dev, "iteration %d of Core En/ROM load fail:%d\n",
+				i, tag);
 			ret = tag;
-			goto irq_err;
+			continue;
 		}
-	}
 
-	/* init for booting wait */
-	init_waitqueue_head(&sdev->boot_wait);
-	sdev->boot_complete = false;
+		/* at this point DSP ROM has been initialized and
+		 * should be ready for code loading and firmware boot
+		 */
+		ret = cl_copy_fw(sdev, tag);
+		if (ret < 0) {
+			dev_err(sdev->dev, "error: iteration %d of load fw failed err: %d\n",
+				i, ret);
+			continue;
+		}
 
-	/* at this point DSP ROM has been initialized and should be ready for
-	 * code loading and firmware boot
-	 */
-	ret = cl_copy_fw(sdev, tag);
-	if (ret < 0) {
-		dev_err(sdev->dev, "error: load fw failed err: %d\n", ret);
-		goto irq_err;
-	}
+		/* init for booting wait */
+		init_waitqueue_head(&sdev->boot_wait);
```

Comment:
can this be moved out of the loop?

Reply:
@plbossart  sure, I"ll take this out of the loop.

Reply:
@ranj063 can you provide this single update, i'd like to merge this before tomorrow. Thanks